stan/math/gpu/matrix_gpu.hpp

```diff
@@ -0,0 +1,283 @@
+#ifndef STAN_MATH_GPU_MATRIX_GPU_HPP
+#define STAN_MATH_GPU_MATRIX_GPU_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/opencl_context.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/arr/fun/value_of.hpp>
+#include <stan/math/prim/mat/err/check_matching_dims.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <CL/cl.hpp>
+#include <iostream>
+#include <string>
+#include <vector>
+/*
+ *  @file stan/math/gpu/matrix_gpu.hpp
+ *  @brief The matrix_gpu class - allocates memory space on the GPU,
+ *    functions for transfering matrices to and from the GPU
+ */
+
+namespace stan {
+namespace math {
+
+/**
+ * This class represents a matrix on the GPU.
+ *
+ * The matrix data is stored in the oclBuffer_.
+ *
+ */
+class matrix_gpu {
+ private:
+  /** cl::Buffer provides functionality for working with OpenCL buffer.
+   * An OpenCL buffer allocates the memory in the device that
+   * is provided by the context.
+   */
+  cl::Buffer oclBuffer_;
+  const int rows_;
+  const int cols_;
+
+ public:
+  int rows() const { return rows_; }
+
+  int cols() const { return cols_; }
+
+  int size() const { return rows_ * cols_; }
+
+  const cl::Buffer& buffer() const { return oclBuffer_; }
+
+  matrix_gpu() : rows_(0), cols_(0) {}
+
+  matrix_gpu(const matrix_gpu& a) : rows_(a.rows()), cols_(a.cols()) {
+    // this check is needed because creating OpenCL
```

Comment:
I don't understand this comment. What does this mean? What creation are you referring to?

Reply:
It's a two line comment
```
    // this check is needed because creating OpenCL
    // buffers of size 0 throws an OpenCL exception
```
OpenCL errors if you try to make a buffer of size 0